File: akka-essentials/Chapter02/FirstAkkaApplicationScala/src/main/scala/akka/first/app/mapreduce/actors/AggregateActor.scala
package akka.first.app.mapreduce.actors

import akka.actor.Actor
import akka.first.app.mapreduce.{ReduceData, Result}
import scala.collection.immutable.Map
import scala.collection.mutable.HashMap

class AggregateActor extends Actor {
  val finalReducedMap = new HashMap[String, Int]

  def receive: Receive = {
    case ReduceData(reduceDataMap) =>
      aggregateInMemoryReduce(reduceDataMap)
    case Result =>
      sender() ! finalReducedMap.toString()
  }

  def aggregateInMemoryReduce(reducedList: Map[String, Int]): Unit = {
    for ((key,value) <- reducedList) {
      if (finalReducedMap contains key)
        finalReducedMap(key) = (value + finalReducedMap.get(key).get)
      else
        finalReducedMap += (key -> value)
    }
  }

}

File: akka-essentials/Chapter02/FirstAkkaApplication2/src/main/java/akka/first/app/mapreduce/actors/AggregateActor.java
package akka.first.app.mapreduce.actors;

import akka.actor.AbstractActor;
import akka.first.app.mapreduce.messages.*;
import java.util.HashMap;
import java.util.Map;

/**
 * The Aggregate actor receives the reduced data list from the Master actor
 * and aggregates it into one big list.
 */
public class AggregateActor extends AbstractActor {
    private Map<String, Integer> finalReducedMap =
        new HashMap<String, Integer>();

    public Receive createReceive() {
        return receiveBuilder()
            .match(ReduceData.class, reduceData -> {
                aggregateInMemoryReduce(reduceData.getReduceDataList());
            })
            .match(Result.class, result -> {
                getSender().tell(finalReducedMap.toString(), getSelf());
            })
            .match(Object.class, message -> {
                unhandled(message);
            })
            .build();
    }

    private void aggregateInMemoryReduce(Map<String, Integer> reducedList) {
        Integer count = null;
        for (String key : reducedList.keySet()) {
            if (finalReducedMap.containsKey(key)) {
                count = reducedList.get(key) + finalReducedMap.get(key);
                finalReducedMap.put(key, count);
            } else {
                finalReducedMap.put(key, reducedList.get(key));
            }
        }
    }

}
